### PR TITLE
Plots and fixes for benchmarks

### PR DESCRIPTION
**Pull request by hennyg888**
Plots were added for the following benchmarking scripts, so as to present the most important data found in the original output tables in a visual format that is easier to analyze:
```
benchmark_shallow_water_model.jl
benchmark_incompressible_model.jl
strong_scaling_shallow_water_model.jl
weak_scaling_shallow_water_model.jl
strong_scaling_incompressible_model.jl
```
The plotting requires the Plots and PyPlot packages.

Additional fixes were also made.

Three manual calls to the garbage collection function, `GC()` was added in `src/Benchmarks.jl` which enabled `benchmark_shallow_water_model.jl` and `benchmark_incompressible_model.jl` to run multiple benchmarking cases with large grids without running out of memory. One such call is `GC( true)` which triggers a "full" garbage collection that is able to remove younger unreferenced objects.

For the _single files of the scaling benchmarking scripts, the `@benchmark` portion was adjusted to have `evals=1` on top of the original configuration of `samples=10`. This allowed larger ranks of CPUs to be benchmarked. Originally, without the configuration limiting evaluations to 1, as the number of CPUs increased and the time per evaluation decreased, the number of times the benchmarked bit of code was run became too great and caused deadlocks between the MPI processes.

### CODE REVIEW

**Comment by hennyg888** [comment]
@francispoulin @ali-ramadhan if it is convenient for you, please review my pull request. As I am not a maintainer or collaborator, I am unable to formally request reviewers, so I had to do it through this comment.

**Review by navidcy — COMMENTED** [review]
Is PyPlot necessary? Doesn’t Plots.jl has all functionality needed?

File: benchmark/Project.toml
Comment (navidcy) [inline benchmark/Project.toml:13]: Is Oceananigans supposed to be here? If not how come things were working before? Or was it broken?

**Comment by navidcy** [comment]
Btw @hennyg888, before any comment I should have mentioned: welcome to Oceananigans.jl! 🙏

**Comment by hennyg888** [comment]
@navidcy PyPlot is used for the log2 scale for the x-axis of certain graphs. The normal backend of Plots,jl only has natural log scale axis option. Log-base-2 scale works best for these plots since most of them have an independent variable that scales based on powers of 2.

The inclusion of Oceananigans in `benchmark/Project.toml` is likely due to a Julia package config suggested by @ali-ramadhan. What was done was that in Julia's package manager, `dev ..` was inputted to enable what I assume is the dev version of the Oceananigans package folder. Before this was done, there were numerous output errors with the benchmark scripts, and doing so fixes them all.

File: benchmark/Project.toml
Comment (glwagner) [reply benchmark/Project.toml:13]: I don't think we want Oceananigans here... we want to use 

```julia
push!(LOAD_PATH, joinpath(@__DIR__, ".."))
```

right at the top of the script (just like the docs):

https://github.com/CliMA/Oceananigans.jl/blob/29bdcb32cda5071ad981ddd4e97d22d02ced8001/docs/make.jl#L1

This is a convenient pattern for getting the benchmarks to run with the "local" Oceananigans code, while having a separate `Project.toml` for the benchmarks.

**Comment by glwagner** [comment]
I think what you likely want to do is to put

```julia
push!(LOAD_PATH, joinpath(@__DIR__, ".."))
```

at the top of every script in `benchmarks/`, mirroring what we do with the docs.

Using `dev ..` will also work, but as you've noted is a bit inconvenient. Also it leads to annoying issues like the present one where `Oceananigans` is included in the `Project.toml` (this will prevent others from running benchmarks on their "local" Oceananigans code).

I also recommend using `git diff file_name.jl` before doing `git add` so that you can ensure the code you're adding is intentional.

File: benchmark/Project.toml
Comment (hennyg888) [reply benchmark/Project.toml:13]: @glwagner does anything in Manifest.toml need to be changed as well?

**Comment by hennyg888** [comment]
@glwagner I've put in the fixes you requested. I've added `push!(LOAD_PATH, joinpath(@__DIR__, ".."))` to the top of every script in the benchmarks folder. I've also deleted Oceananigans from Project.toml.
Before I push my local commits, does Manifest.toml require any changes?

**Comment by glwagner** [comment]
> Before I push my local commits, does Manifest.toml require any changes?

`Manifest.toml` typically has to be synchronized with `Project.toml`.

In a perfect world (if our package compatibility bounds in `Project.toml` were perfect) then users could generate `Manifest.toml` on their local machine without requiring a committed version.

However, it turns out to be a bit annoying to keep compatibility bounds up to date, so we use the `Manifest.toml` as a hack. Since we can't really review changes to the `Manifest.toml` it's best to be conservative about updating packages, etc.

**Comment by navidcy** [comment]
Also, this PR shows that tests have passed. But do the tests run the benchmarks? @ali-ramadhan is there a benchmark pipeline and how can we trigger it if so?

**Comment by tomchor** [comment]
Restarting the tests since we were having problems with the servers these past couple of days.

**Review by ali-ramadhan — APPROVED** [review]
This is great work, thanks @hennyg888! We should be able to continue building on top of this PR so future benchmarks can also output useful plots. Also added you as a repo collaborator although all PRs still require at least one approval to merge.

@navidcy Unfortunately there is no benchmarking CI pipeline right now :( Always hoped to add one but never got around to it. I think it's a bit tough to do correctly. It should be easy to run the benchmark scripts to make sure they run without errors but they still take up a lot of memory.

Getting accurate benchmarks from CI would be great and super useful but would probably be trickier as many benchmarks completely saturate the GPU so you can only run one at a time. So probably we need to run the benchmark scripts one by one on a server with nothing else going on so we might need more resources as right now the CI servers are shared with other users/jobs. We could also run them nightly or weekly like the verification experiments CI pipeline and set up alerts for performance regressions.

PS: Since benchmarks aren't actually tested right now and can't affect CI this PR could have been merged without CI passing, but definitely better if CI does pass.

**Pull request merged**
